```
x86_64-unknown-linux-gnu-nolibc-gcc> ../../gcc-13.3.0/gcc/analyzer/diagnostic-manager.cc: In member function 'void ana::saved_diagnostic::dump_as_dot_node(pretty_printer*) const':
x86_64-unknown-linux-gnu-nolibc-gcc> ../../gcc-13.3.0/gcc/analyzer/diagnostic-manager.cc:802:28: warning: unknown conversion type character 'E' in format [8;;https://gcc.gnu.org/onlinedocs/gcc/Warning-Options.html#index-Wformat=-Wformat=8;;]
x86_64-unknown-linux-gnu-nolibc-gcc>   802 |     pp_printf (pp, "var: %qE\n", m_var);
x86_64-unknown-linux-gnu-nolibc-gcc>       |                            ^
x86_64-unknown-linux-gnu-nolibc-gcc> ../../gcc-13.3.0/gcc/analyzer/diagnostic-manager.cc:802:20: warning: too many arguments for format [8;;https://gcc.gnu.org/onlinedocs/gcc/Warning-Options.html#index-Wformat-extra-args-Wformat-extra-args8;;]
x86_64-unknown-linux-gnu-nolibc-gcc>   802 |     pp_printf (pp, "var: %qE\n", m_var);
x86_64-unknown-linux-gnu-nolibc-gcc>       |                    ^~~~~~~~~~~~
error: build of '/nix/store/yr237v031vhn8qwzivg4p2mjh6581nyk-llvm-18.1.8.drv' on 'ssh-ng://nixbld@fd7a:115c:a1e0::22' failed: builder for '/nix/store/yr237v031vhn8qwzivg4p2mjh6581nyk-llvm-18.1.8.drv' failed with exit code 1;
       last 25 log lines:
       > 0x87fcf9 cp_parser_declaration_seq_opt(cp_parser*)
       >         ???:0
       > 0x880196 cp_parser_namespace_definition(cp_parser*)
       >     ???:0
       > 0x880b44 cp_parser_declaration(cp_parser*, tree_node*)
       >  ???:0
       > 0x8812d9 c_parse_file()
       >         ???:0
       > 0x977021 c_common_parse_file()
       >  ???:0
       > Please submit a full bug report, with preprocessed source (by using -freport-bug).
       > Please include the complete backtrace with any bug report.
       > See <https://gcc.gnu.org/bugs/> for instructions.
       > [2359/4385] Building CXX object lib/Target/AArch64/CMakeFiles/LLVMAArch64CodeGen.dir/AArch64A57FPLoadBalancing.cpp.o
       > [2360/4385] Building CXX object lib/Target/AArch64/CMakeFiles/LLVMAArch64CodeGen.dir/GISel/AArch64PreLegalizerCombiner.cpp.o
       > [2361/4385] Building CXX object lib/Target/AArch64/CMakeFiles/LLVMAArch64CodeGen.dir/GISel/AArch64PostLegalizerCombiner.cpp.o
```

--- a/pkgs/development/compilers/zig/0.10/default.nix
+++ b/pkgs/development/compilers/zig/0.10/default.nix
@@ -89,9 +89,8 @@ stdenv.mkDerivation (finalAttrs: {
 
   passthru = {
     hook = callPackage ./hook.nix { zig = finalAttrs.finalPackage; };
-    cc = callPackage ../cc.nix {
-      zig = finalAttrs.finalPackage;
-    };
+    cc = callPackage ../cc.nix { zig = finalAttrs.finalPackage; };
+    stdenv = callPackage ../stdenv.nix { zig = finalAttrs.finalPackage; };
     tests = {
       version = testers.testVersion {
         package = finalAttrs.finalPackage;

--- a/pkgs/development/compilers/zig/0.10/default.nix
+++ b/pkgs/development/compilers/zig/0.10/default.nix
@@ -89,6 +89,9 @@ stdenv.mkDerivation (finalAttrs: {
 
   passthru = {
     hook = callPackage ./hook.nix { zig = finalAttrs.finalPackage; };
+    cc = callPackage ../cc.nix {
+      zig = finalAttrs.finalPackage;
+    };
     tests = {
       version = testers.testVersion {
         package = finalAttrs.finalPackage;

--- a/pkgs/development/compilers/zig/0.11/default.nix
+++ b/pkgs/development/compilers/zig/0.11/default.nix
@@ -83,9 +83,8 @@ stdenv.mkDerivation (finalAttrs: {
 
   passthru = {
     hook = callPackage ./hook.nix { zig = finalAttrs.finalPackage; };
-    cc = callPackage ../cc.nix {
-      zig = finalAttrs.finalPackage;
-    };
+    cc = callPackage ../cc.nix { zig = finalAttrs.finalPackage; };
+    stdenv = callPackage ../stdenv.nix { zig = finalAttrs.finalPackage; };
     tests = {
       version = testers.testVersion {
         package = finalAttrs.finalPackage;

--- a/pkgs/development/compilers/zig/0.11/default.nix
+++ b/pkgs/development/compilers/zig/0.11/default.nix
@@ -83,6 +83,9 @@ stdenv.mkDerivation (finalAttrs: {
 
   passthru = {
     hook = callPackage ./hook.nix { zig = finalAttrs.finalPackage; };
+    cc = callPackage ../cc.nix {
+      zig = finalAttrs.finalPackage;
+    };
     tests = {
       version = testers.testVersion {
         package = finalAttrs.finalPackage;

--- a/pkgs/development/compilers/zig/0.12/default.nix
+++ b/pkgs/development/compilers/zig/0.12/default.nix
@@ -95,6 +95,9 @@ stdenv.mkDerivation (finalAttrs: {
 
   passthru = {
     hook = callPackage ./hook.nix { zig = finalAttrs.finalPackage; };
+    cc = callPackage ../cc.nix {
+      zig = finalAttrs.finalPackage;
+    };
     tests = {
       version = testers.testVersion {
         package = finalAttrs.finalPackage;

--- a/pkgs/development/compilers/zig/0.12/default.nix
+++ b/pkgs/development/compilers/zig/0.12/default.nix
@@ -95,9 +95,8 @@ stdenv.mkDerivation (finalAttrs: {
 
   passthru = {
     hook = callPackage ./hook.nix { zig = finalAttrs.finalPackage; };
-    cc = callPackage ../cc.nix {
-      zig = finalAttrs.finalPackage;
-    };
+    cc = callPackage ../cc.nix { zig = finalAttrs.finalPackage; };
+    stdenv = callPackage ../stdenv.nix { zig = finalAttrs.finalPackage; };
     tests = {
       version = testers.testVersion {
         package = finalAttrs.finalPackage;

--- a/pkgs/development/compilers/zig/0.13/default.nix
+++ b/pkgs/development/compilers/zig/0.13/default.nix
@@ -95,6 +95,8 @@ stdenv.mkDerivation (finalAttrs: {
 
   passthru = {
     hook = callPackage ./hook.nix { zig = finalAttrs.finalPackage; };
+    cc = callPackage ../cc.nix { zig = finalAttrs.finalPackage; };
+    stdenv = callPackage ../stdenv.nix { zig = finalAttrs.finalPackage; };
     tests = {
       version = testers.testVersion {
         package = finalAttrs.finalPackage;

--- a/pkgs/development/compilers/zig/0.9/default.nix
+++ b/pkgs/development/compilers/zig/0.9/default.nix
@@ -91,9 +91,8 @@ stdenv.mkDerivation (finalAttrs: {
 
   passthru = {
     hook = callPackage ./hook.nix { zig = finalAttrs.finalPackage; };
-    cc = callPackage ../cc.nix {
-      zig = finalAttrs.finalPackage;
-    };
+    cc = callPackage ../cc.nix { zig = finalAttrs.finalPackage; };
+    stdenv = callPackage ../stdenv.nix { zig = finalAttrs.finalPackage; };
     tests = {
       version = testers.testVersion {
         package = finalAttrs.finalPackage;

--- a/pkgs/development/compilers/zig/0.9/default.nix
+++ b/pkgs/development/compilers/zig/0.9/default.nix
@@ -91,6 +91,9 @@ stdenv.mkDerivation (finalAttrs: {
 
   passthru = {
     hook = callPackage ./hook.nix { zig = finalAttrs.finalPackage; };
+    cc = callPackage ../cc.nix {
+      zig = finalAttrs.finalPackage;
+    };
     tests = {
       version = testers.testVersion {
         package = finalAttrs.finalPackage;

--- a/pkgs/development/compilers/zig/cc.nix
+++ b/pkgs/development/compilers/zig/cc.nix
@@ -1,0 +1,42 @@
+{
+  lib,
+  wrapCCWith,
+  makeWrapper,
+  runCommand,
+  targetPlatform,
+  targetPackages,
+  zig,
+}:
+wrapCCWith {
+  cc =
+    runCommand "zig-cc-${zig.version}"
+      {
+        pname = "zig-cc";
+        inherit (zig) version meta;
+
+        nativeBuildInputs = [ makeWrapper ];
+
+        passthru.isZig = true;
+        inherit zig;
+      }
+      ''
+        mkdir -p $out/bin
+        for tool in ar cc c++ objcopy; do
+          makeWrapper "$zig/bin/zig" "$out/bin/$tool" \
+            --add-flags "$tool" \
+            --run "export ZIG_GLOBAL_CACHE_DIR=\$(mktemp -d)"
+        done
+
+        mv $out/bin/c++ $out/bin/clang++
+        mv $out/bin/cc $out/bin/clang
+      '';
+
+  nixSupport.cc-cflags =
+    [
+      "-target"
+      "${targetPlatform.parsed.cpu.name}-${targetPlatform.parsed.kernel.name}-${targetPlatform.parsed.abi.name}"
+    ]
+    ++ lib.optional (
+      targetPlatform.isLinux && !(targetPackages.isStatic or false)
+    ) "-Wl,-dynamic-linker=${targetPackages.stdenv.cc.bintools.dynamicLinker}";
+}

--- a/pkgs/development/compilers/zig/generic.nix
+++ b/pkgs/development/compilers/zig/generic.nix
@@ -71,6 +71,9 @@ stdenv.mkDerivation (finalAttrs: {
     hook = callPackage ./hook.nix {
       zig = finalAttrs.finalPackage;
     };
+    cc = callPackage ./cc.nix {
+      zig = finalAttrs.finalPackage;
+    };
   };
 
   meta = {

--- a/pkgs/development/compilers/zig/generic.nix
+++ b/pkgs/development/compilers/zig/generic.nix
@@ -74,6 +74,9 @@ stdenv.mkDerivation (finalAttrs: {
     cc = callPackage ./cc.nix {
       zig = finalAttrs.finalPackage;
     };
+    stdenv = callPackage ./stdenv.nix {
+      zig = finalAttrs.finalPackage;
+    };
   };
 
   meta = {

--- a/pkgs/development/compilers/zig/stdenv.nix
+++ b/pkgs/development/compilers/zig/stdenv.nix
@@ -1,0 +1,6 @@
+{
+  stdenv,
+  overrideCC,
+  zig,
+}:
+overrideCC stdenv zig.cc

--- a/pkgs/stdenv/cross/default.nix
+++ b/pkgs/stdenv/cross/default.nix
@@ -75,6 +75,8 @@ in lib.init bootStages ++ [
                then buildPackages.llvmPackages.libcxxClang
              else if crossSystem.useLLVM or false
                then buildPackages.llvmPackages.clang
+             else if crossSystem.useZig or false
+               then buildPackages.zig.cc
              else buildPackages.gcc;
 
         extraNativeBuildInputs = old.extraNativeBuildInputs

--- a/pkgs/top-level/all-packages.nix
+++ b/pkgs/top-level/all-packages.nix
@@ -24696,6 +24696,8 @@ with pkgs;
   };
   zig = zig_0_13;
 
+  zigStdenv = if stdenv.cc.isZig then stdenv else lowPrio zig.passthru.stdenv;
+
   zimlib = callPackage ../development/libraries/zimlib { };
 
   zita-convolver = callPackage ../development/libraries/audio/zita-convolver { };

